### DISCUSSION TITLE
HttpSensor for CVAT Webhooks

### DISCUSSION THREAD

**Discussion by jaimecorton** [apache/airflow#35749] [Q&A]
I am trying to configure the activation of a Python script through a sensor from an Airflow DAG that senses the JSON sended by the CVAT webhook once certain conditions are met (for example when the annotation task has finished). I deployed CVAT and Airflow with Docker in 2 separate machines with the same OS and Kernel (Operating System: Ubuntu 22.04.3 LTS, Kernel: Linux 5.15.0-88-generic, Architecture: x86-64).

I already have a working url for the CVAT Webhook, lets call that url https://direction.com/login/, so far, while trying the Ping option from the CVAT UI the response is correct (200) and I can see the produced JSON through ~$ docker logs {container id} comand. I also created an Airflow connection to https://direction.com/ .Moreover, on my Airflow DAG (that doesn't have any import error or anything like that) I am using an [httpSensor](https://airflow.apache.org/docs/apache-airflow-providers-http/stable/_api/airflow/providers/http/sensors/http/index.html) with the corresponding http_conn_id and the endpoint /login as it should be.

Once all of this is done, the expected behaviour should be that once the DAG's http sensor reads the webhook's JSON message on the specifyied URL the script execution should be activated.

Here is the code that I have for the airflow DAG:

```
@dag(
default_args={
    'owner': 'OWNER',
    'start_date': datetime(2023, 1, 1),
},
schedule_interval=None, 
tags=['cvat_dag','webhook'],
catchup=False,)

def cvat_webhook_dag():

    @task
    def http_sensor_task():
            return HttpSensor(
                task_id='cvat_webhook_sensor',
                http_conn_id='cvat_webhook',
                endpoint='/login/',
                method='GET',
                response_check=lambda response: "CVAT_trial_task_1" in json.dumps(response.json()), 
                timeout=120,
                poke_interval=60
            )
    
    @task
    def execute_script():
        # Execute your script here
        print('execute script triggered')
    
    @task
    def restart_dag_fun():
        restart_dag = TriggerDagRunOperator(
            task_id = 'restart_dag',
            trigger_dag_id = 'cvat_webhook_sensor',
        )

    # Set the task dependencies
    http_sensor_task() >> execute_script() >> restart_dag_fun()
```

I also changed the HttpSensor so that response_check=lambda response: response.status_code==200 to force the script execution with independence of the JSON content, but It didn't work neither. I checked the logs from Airflow and they don't give any meaningfull informationit just says that the execution failed.

Any suggestion would be great. I am not very familiar with Airflow so probably something is wrong but I don't know what might be the solution. There's a discussion about it as well in Stack Overflow (https://stackoverflow.com/questions/77497014/cvat-webhooks-airflow-dag-trigger)

Thank you in advance for your help.

**Comment by potiuk** [comment]
There might be many reasons and you can simply debug the code.

 Look at the sensor code in Airdlow repo (it's open-source) and see if more debugging logging options are available. - if not - the simplest way you can copy the HttpSensor code to your DAG and run it instead - then you can add any debugging code you need - this is Python, you already program your DAg in Python so copying a few lines of code to make it MyHttpSenso and using it instead  (and adding debug code).should be easy. 

You should see which http provider version you have (from provider package version in 'pip' or providers tab in Airflow UI or 'airflow providers' CLI) and the sensor code will be in the corresponding tag providers-http/VERSION in airflow/providers/http in airflow repo

**Reply by potiuk** [reply]
Sorry for typos -autocomplete :)

**Reply by jaimecorton** [reply]
Hi,

Thanks for your quick answer, The Airflow UI http provider version is 3.0.0 and the Airflow HttpSensor http provider (as seen in the repo [https://github.com/apache/airflow/blob/main/airflow/providers/http/__init__.py](https://github.com/apache/airflow/blob/main/airflow/providers/http/__init__.py))  is 4.7.0 would that be problematic? Or is there any kind of forward/backward compatibility? 
I think as far as the Airflow UI http provider version is more than 2.5.0 there  shouldn't be a problem, or am I wrong?

```

try:
    from airflow import __version__ as airflow_version
except ImportError:
    from airflow.version import version as airflow_version

if packaging.version.parse(packaging.version.parse(airflow_version).base_version) < packaging.version.parse(
    "2.5.0"
):
    raise RuntimeError(
        f"The package `apache-airflow-providers-http:{__version__}` needs Apache Airflow 2.5.0+"
    )
```

**Reply by potiuk** [reply]
Generally speaking you are free to upgrade to newer versions of provider at any time. Unless there is a problem that either prevent you from upgrading (dependencies) or some breaking change you would not want to use.

I think you might look at the release notes. The system is designed to self-serve and self-debug. There is no way any single person can know all the 3000+ operators we have. So if there is something changed betwen those versions - that would be writen in release notes. This would not be scalable if handful of maintainers would have to answer question about every single providers upgrade and be authoritative about it. 

And also don't say my word for granted - it **might** be that it will not work even if I tell you it should - this is ultimately your responsibility to check as this software (according to licence) comes with no guarantees whatsoever (it's developed for free by volunteers). Also this is the reason why we expect our users who - out-of-necessity - are Python programers to be able to debug and test it on their own.

But yes. INTENTION is that it should work,
